vlrouter.go

```diff
@@ -0,0 +1,662 @@
+	client := api.NewGobgpApiClient(conn)
+	if client == nil {
+		log.Errorf("Gobgpapi stream invalid")
+		return nil
+	}
+
+	stream, err := client.ModPath(context.Background())
+	log.Infof("The stream is ", stream)
+	if err != nil {
+		log.Errorf("Fail to enforce Modpathi", err)
+		return err
+	}
+	err = stream.Send(arg)
+	if err != nil {
+		log.Errorf("Failed to send strean", err)
+		return err
+	}
+	stream.CloseSend()
+	res, e := stream.CloseAndRecv()
+	if e != nil {
+		log.Errorf("Falied toclose stream ")
+		return e
+	}
+	if res.Code != api.Error_SUCCESS {
+		return fmt.Errorf("error: code: %d, msg: %s", res.Code, res.Msg)
+	}
+	return nil
+}
+
+// Add a vlan.
+// This is mainly used for mapping vlan id to Vxlan VNI
+func (self *Vlrouter) AddVlan(vlanId uint16, vni uint32) error {
+	// FIXME: Add this for multiple VRF support
+	return nil
+}
+
+// Remove a vlan
+func (self *Vlrouter) RemoveVlan(vlanId uint16, vni uint32) error {
+	// FIXME: Add this for multiple VRF support
+	return nil
+}
+
+/* AddEndpoint does the following :
+1)Adds a remote endpoint and associated flows to OVS
+2)The remotes routes can be 3 endpoint types :
+  a) internal - json rpc based learning from peer netplugins/ofnetagents in the cluster
+	b) external - remote endpoint learn via BGP
+	c) external-bgp - endpoint of BGP peer
+*/
+func (self *Vlrouter) AddEndpoint(endpoint *OfnetEndpoint) error {
+
+	nexthopEp := self.agent.getEndpointByIp(net.ParseIP(self.MyBgpPeer))
+	if nexthopEp != nil && nexthopEp.PortNo != 0 {
+		endpoint.MacAddrStr = nexthopEp.MacAddrStr
+		endpoint.PortNo = nexthopEp.PortNo
+	} else {
+		endpoint.PortNo = 0
+		endpoint.MacAddrStr = " "
+		if endpoint.EndpointType != "external-bgp" {
+			//for the remote endpoints maintain a cache of
+			//routes that need to be resolved to next hop.
+			// bgp peer resolution happens via ARP and hence not
+			//maintainer in cache.
+			log.Info("Putting in endpoint info to cache")
+			self.unresolvedEPs.PushBack(endpoint.EndpointID)
+			log.Info(self.unresolvedEPs)
+		}
+	}
+	if endpoint.EndpointType == "external-bgp" {
+		self.MyBgpPeer = endpoint.IpAddr.String()
+	}
+	log.Infof("AddEndpoint call for endpoint: %+v", endpoint)
+	//Install a flow entry for vlan mapping and point it to IP table
+	// FIXME: Remove this. This has to be added as a part of Add vlan .
+	portVlanFlow, err := self.vlanTable.NewFlow(ofctrl.FlowMatch{
+		Priority:  FLOW_MATCH_PRIORITY,
+		InputPort: endpoint.PortNo,
+	})
+
+	// Set the vlan and install it
+	// FIXME: Dont set the vlan till multi-vrf support. We cant pop vlan unless flow matches on vlan
+	// portVlanFlow.SetVlan(endpoint.Vlan)
+	if err == nil {
+		portVlanFlow.Next(self.ipTable)
+		// save the flow entry
+		self.portVlanFlowDb[endpoint.PortNo] = portVlanFlow
+	} else {
+		portVlanFlow = self.portVlanFlowDb[endpoint.PortNo]
+		portVlanFlow.Next(self.ipTable)
+	}
+	fmt.Println(endpoint.PortNo)
+	outPort, err := self.ofSwitch.OutputPort(endpoint.PortNo)
```

Comment:
Not really . We let the caller of the AddEndpoint to pass on the output port. This is cases where uplink is down, bgp peer is down the portno will be 0 . And it will be resolved if and only when bgp peer is up and connected.